python/shotgun_desktop/startup.py

```diff
@@ -478,276 +334,156 @@ def __extract_command_line_argument(arg_name):
+    )
+    mgr.pipeline_configuration = pc["id"]
+
+    def pre_engine_start_callback(tk, ctx):
+        # If the pipeline configuration found in Shotgun doesn't match what we have locally, we have a
+        # problem.
+        if pc["id"] != tk.pipeline_configuration.get_shotgun_id():
+            raise InvalidPipelineConfiguration(pc, tk.pipeline_configuration)
+
+        # If the pipeline configuration we got from Shotgun is not assigned to a project, we might have
+        # some patching to be done to local site configuration.
+        if pc["project"] is None:
+
+            # make sure that the version of core we are using supports the new-style site configuration
+            if not __supports_pipeline_configuration_upgrade(tk.pipeline_configuration):
+                raise UpgradeCoreError(
+                    "Running a site configuration without the Template Project requires core v0.16.8 "
+                    "or higher.",
+                    pc_path
+                )
 
-    if not __toolkit_supports_authentication_module(sgtk):
-        raise UpgradeCoreError(
-            "This version of the Shotgun Desktop only supports core 0.16.4 and higher.",
-            default_site_config
-        )
-    # initialize the tk-desktop engine for an empty context
-    splash.set_message("Starting desktop engine.")
-    app.processEvents()
+            # If the configuration on disk is not the site configuration, update it to the site config.
+            if not tk.pipeline_configuration.is_site_configuration():
+                tk.pipeline_configuration.convert_to_site_config()
+
+        splash.set_message("Launching Engine...")
 
-    ctx = tk.context_empty()
-    engine = sgtk.platform.start_engine("tk-desktop", tk, ctx)
+    # We need to validate a few things before the engine starts.
+    mgr.pre_engine_start_callback = pre_engine_start_callback
+
+    engine = mgr.bootstrap_engine("tk-desktop")
 
     if not __desktop_engine_supports_authentication_module(engine):
         raise UpgradeEngineError(
             "This version of the Shotgun Desktop only supports tk-desktop engine 2.0.0 and higher.",
-            default_site_config
+            pc_path
         )
 
+    return engine
+
+
+def __start_engine_in_zero_config(app, splash, user):
+    """
+    Launch into the engine using the new zero config based bootstrap.
+
+    :param app: Application object for event processing.
+    :param splash: Splash dialog to update user on what is currently going on
+    :param user: Current ShotgunUser.
+
+    :returns: Toolkit engine that was started.
+    """
+    # The startup is up to date, now it's time to bootstrap Toolkit.
+    import sgtk
+    mgr = sgtk.bootstrap.ToolkitManager(user)
+
+    # Allows to take over the site config to use with Desktop without impacting the projects
+    # configurations.
+    mgr.base_configuration = os.environ.get(
+        "SHOTGUN_DESKTOP_CONFIG_FALLBACK_DESCRIPTOR",
```

Comment:
Nope, this is for testing. We can discuss it tomorrow if you want. Its tricky to explain in text.